# ANDROID APPLICATION DATA ANALYSIS


Dataset: https://drive.google.com/file/d/1lGIrooUAGjgNSSfD2C9VrS4Oy3paLTdO/view?usp=sharing

* importing important libraries

In [180]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import seaborn as sns

In [181]:
df = pd.read_csv(r'C:\Users\shoba\Desktop\data_science\android_application\Mobile_Applications.csv')

In [182]:
df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

In [183]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


* rating , type , content rating, current version and android version has null values
* except rating all are having object data type 

# EDA

In [184]:
df.App.unique()

array(['Photo Editor & Candy Camera & Grid & ScrapBook',
       'Coloring book moana',
       'U Launcher Lite – FREE Live Cool Themes, Hide Apps', ...,
       'Parkinson Exercices FR', 'The SCP Foundation DB fr nn5n',
       'iHoroscope - 2018 Daily Horoscope & Astrology'], dtype=object)

In [185]:
df.App.nunique()

9660

* we can't convert this much unique value to numeric 

In [186]:
df.Category.unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)

In [187]:
df.Category.nunique()

34

In [188]:
df.Category.value_counts()

FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
LIBRARIES_AND_DEMO       85
AUTO_AND_VEHICLES        85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
PARENTING                60
COMICS                   60
BEAUTY                   53
1.9                       1
Name: Category, dtype: int64

* we can see here that it has inopropriate data "1.9"
* and we can change the these values to numeric as we have only 34 unique values
* we can even create a seperate dataframe containg the App respective of the category
* and we can drop the App columns

In [189]:
df.Rating.isnull().sum()

1474

In [190]:
df.Rating.value_counts()

4.4     1109
4.3     1076
4.5     1038
4.2      952
4.6      823
4.1      708
4.0      568
4.7      499
3.9      386
3.8      303
5.0      274
3.7      239
4.8      234
3.6      174
3.5      163
3.4      128
3.3      102
4.9       87
3.0       83
3.1       69
3.2       64
2.9       45
2.8       42
2.7       25
2.6       25
2.5       21
2.3       20
2.4       19
1.0       16
2.2       14
1.9       13
2.0       12
1.7        8
1.8        8
2.1        8
1.6        4
1.4        3
1.5        3
1.2        1
19.0       1
Name: Rating, dtype: int64

In [191]:
df.Rating.describe()

count    9367.000000
mean        4.193338
std         0.537431
min         1.000000
25%         4.000000
50%         4.300000
75%         4.500000
max        19.000000
Name: Rating, dtype: float64

* we have null vailues in the rating colums
* it seems we have a outlier 19 in the ratings column

In [192]:
df[df['Rating'] == 19]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

* it seems the datapoint has shhifted one column before from the category
* see it's also contains rating in the category.
* we have to shift the values one step behind from category we should not shift App value

In [193]:
df.iloc[10472] = df.iloc[10472].shift()
# shiftng values to the right side of the datapoint

In [194]:
df.iloc[10472]

App                                                   NaN
Category          Life Made WI-Fi Touchscreen Photo Frame
Rating                                                1.9
Reviews                                              19.0
Size                                                 3.0M
Installs                                           1,000+
Type                                                 Free
Price                                                   0
Content Rating                                   Everyone
Genres                                                NaN
Last Updated                            February 11, 2018
Current Ver                                        1.0.19
Android Ver                                    4.0 and up
Name: 10472, dtype: object

* the App value also shifted to right

In [195]:
df.App.isna().sum()

1

In [196]:
df.App.fillna("Life Made WI-Fi Touchscreen Photo Frame" , inplace=True)

In [197]:
df.iloc[10472]

App               Life Made WI-Fi Touchscreen Photo Frame
Category          Life Made WI-Fi Touchscreen Photo Frame
Rating                                                1.9
Reviews                                              19.0
Size                                                 3.0M
Installs                                           1,000+
Type                                                 Free
Price                                                   0
Content Rating                                   Everyone
Genres                                                NaN
Last Updated                            February 11, 2018
Current Ver                                        1.0.19
Android Ver                                    4.0 and up
Name: 10472, dtype: object

In [198]:
df.Category.value_counts()

FAMILY                                     1972
GAME                                       1144
TOOLS                                       843
MEDICAL                                     463
BUSINESS                                    460
PRODUCTIVITY                                424
PERSONALIZATION                             392
COMMUNICATION                               387
SPORTS                                      384
LIFESTYLE                                   382
FINANCE                                     366
HEALTH_AND_FITNESS                          341
PHOTOGRAPHY                                 335
SOCIAL                                      295
NEWS_AND_MAGAZINES                          283
SHOPPING                                    260
TRAVEL_AND_LOCAL                            258
DATING                                      234
BOOKS_AND_REFERENCE                         231
VIDEO_PLAYERS                               175
EDUCATION                               

* we have to fill a category value to the data point which suits it

In [199]:
# to find the best category for the null we are checking the other App name containg the same words and storing the category of the same word
s = []                                     # empty list to store the data of the category 
for i in (df.App[10472].split()):          # this function splita the app name into list seperated between the space and i get the value one by one
    for j,k in zip(df.App, df.Category):   # loop for the app name aand the category
        if i in j:                         # condition statement that the seperated value in the app name 
            s.append(str(k))               # if the condition is true then append the category of the patrticular app name
s = pd.DataFrame(s)                        # creating a data frame to use the value_counts in the list

In [200]:
s.value_counts()

PHOTOGRAPHY                                138
FAMILY                                      16
ART_AND_DESIGN                              10
TOOLS                                        7
SHOPPING                                     7
Life Made WI-Fi Touchscreen Photo Frame      6
SOCIAL                                       5
EDUCATION                                    5
PERSONALIZATION                              4
VIDEO_PLAYERS                                4
HEALTH_AND_FITNESS                           3
BEAUTY                                       3
EVENTS                                       3
ENTERTAINMENT                                3
BOOKS_AND_REFERENCE                          3
LIFESTYLE                                    3
GAME                                         2
MEDICAL                                      2
NEWS_AND_MAGAZINES                           2
PRODUCTIVITY                                 2
DATING                                       2
COMICS       

* we can clearly see that photograpy suits its best 

In [201]:
df.Category.replace('Life Made WI-Fi Touchscreen Photo Frame', 'PHOTOGRAPHY', inplace=True)

In [202]:
df.iloc[10472]

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                      PHOTOGRAPHY
Rating                                                1.9
Reviews                                              19.0
Size                                                 3.0M
Installs                                           1,000+
Type                                                 Free
Price                                                   0
Content Rating                                   Everyone
Genres                                                NaN
Last Updated                            February 11, 2018
Current Ver                                        1.0.19
Android Ver                                    4.0 and up
Name: 10472, dtype: object

In [203]:
df.Rating.isna().sum()

1474

* we are having 1474 null values 
* filling the null values with the mean of the Rating column

In [204]:
df.Rating.describe()

count     9367.0
unique      40.0
top          4.4
freq      1109.0
Name: Rating, dtype: float64

In [205]:
df.Rating.dtype

dtype('O')

* after shifting the dtype changed to object we have change that to float

In [206]:
df['Rating'] = df['Rating'].astype(float)

In [207]:
df.Rating.describe()

count    9367.000000
mean        4.191513
std         0.515735
min         1.000000
25%         4.000000
50%         4.300000
75%         4.500000
max         5.000000
Name: Rating, dtype: float64

In [208]:
df.Rating.fillna(df.Rating.mean(), inplace=True)

In [209]:
df.Reviews.value_counts()

0         596
1         272
2         214
3         175
4         137
         ... 
342912      1
4272        1
5517        1
4057        1
398307      1
Name: Reviews, Length: 6002, dtype: int64

In [210]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10841 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10839 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [211]:
df.Reviews.describe()

count     10841
unique     6002
top           0
freq        596
Name: Reviews, dtype: object

In [212]:
df.Reviews = df.Reviews.astype(np.int64)

In [213]:
df.Reviews.describe()

count    1.084100e+04
mean     4.441119e+05
std      2.927629e+06
min      0.000000e+00
25%      3.800000e+01
50%      2.094000e+03
75%      5.476800e+04
max      7.815831e+07
Name: Reviews, dtype: float64

In [214]:
df.Size.unique()

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 'Varies with device', '9.4M', '15M', '10M',
       '1.2M', '26M', '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k',
       '3.6M', '5.7M', '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M',
       '8.9M', '3.9M', '2.9M', '38M', '32M', '5.4M', '18M', '1.1M',
       '2.2M', '4.5M', '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M',
       '7.1M', '3.7M', '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M',
       '4.9M', '9.5M', '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M',
       '4.0M', '2.3M', '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M',
       '23k', '6.5M', '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M',
       '8.3M', '4.3M', '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M',
       '5.1M', '61M', '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M',
     

In [215]:
df.Size.replace("Varies with device", 0, inplace=True)

In [216]:
df.Size.unique()

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 0, '9.4M', '15M', '10M', '1.2M', '26M',
       '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k', '3.6M', '5.7M',
       '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M', '8.9M', '3.9M',
       '2.9M', '38M', '32M', '5.4M', '18M', '1.1M', '2.2M', '4.5M',
       '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M', '7.1M', '3.7M',
       '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M', '4.9M', '9.5M',
       '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M', '4.0M', '2.3M',
       '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M', '23k', '6.5M',
       '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M', '8.3M', '4.3M',
       '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M', '5.1M', '61M',
       '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M', '6.2M', '18k',
       '5

* instead of having a mb or kb we can seperate the value from t

In [217]:
df.Size.value_counts()

0       1695
11M      198
12M      196
14M      194
13M      191
        ... 
430k       1
429k       1
200k       1
460k       1
619k       1
Name: Size, Length: 461, dtype: int64

In [218]:
memory = []
for i in range(len(df.Size)):
     try:
          d = re.search("k|M", str(df.Size[i]))
          memory.append(re.sub("\d|\.","",str(df.Size[i])))
     except:
          memory.append(df.Size[i])

In [219]:
for i in range(len(df.Size)):
    if df.Size[i] == 0:
        pass
    else:
        df.Size[i] = re.sub(".$","",str(df.Size[i]))

C:\Users\shoba\AppData\Local\Temp/ipykernel_5920/3555852622.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.Size[i] = re.sub(".$","",str(df.Size[i]))


In [220]:
df.Size.unique()

array(['19', '14', '8.7', '25', '2.8', '5.6', '29', '33', '3.1', '28',
       '12', '20', '21', '37', '2.7', '5.5', '17', '39', '31', '4.2',
       '7.0', '23', '6.0', '6.1', '4.6', '9.2', '5.2', '11', '24', 0,
       '9.4', '15', '10', '1.2', '26', '8.0', '7.9', '56', '57', '35',
       '54', '201', '3.6', '5.7', '8.6', '2.4', '27', '2.5', '16', '3.4',
       '8.9', '3.9', '2.9', '38', '32', '5.4', '18', '1.1', '2.2', '4.5',
       '9.8', '52', '9.0', '6.7', '30', '2.6', '7.1', '3.7', '22', '7.4',
       '6.4', '3.2', '8.2', '9.9', '4.9', '9.5', '5.0', '5.9', '13', '73',
       '6.8', '3.5', '4.0', '2.3', '7.2', '2.1', '42', '7.3', '9.1', '55',
       '6.5', '1.5', '7.5', '51', '41', '48', '8.5', '46', '8.3', '4.3',
       '4.7', '3.3', '40', '7.8', '8.8', '6.6', '5.1', '61', '66', '79',
       '8.4', '118', '44', '695', '1.6', '6.2', '53', '1.4', '3.0', '5.8',
       '3.8', '9.6', '45', '63', '49', '77', '4.4', '4.8', '70', '6.9',
       '9.3', '10.0', '8.1', '36', '84', '97', '2.0',

In [221]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [222]:
df.insert(5, 'Memory',memory)

In [223]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews Size Memory     Installs  Type Price Content Rating  \
0      159   19      M      10,000+  Free     0       Everyone   
1      967   14      M     500,000+  Free     0       Everyone   
2    87510  8.7      M   5,000,000+  Free     0       Everyone   
3   215644   25      M  50,000,000+  Free     0           Teen   
4      967  2.8      M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [224]:
df.Memory.value_counts()

M    8830
     1695
k     316
Name: Memory, dtype: int64

In [225]:
df.Memory.unique()

array(['M', '', 'k'], dtype=object)

In [226]:
df.Memory.replace(["M", "", 'k'],[2, 0, 1], inplace=True)

In [227]:
df.Memory.dtype

dtype('int64')

In [228]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  object 
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  object 
 7   Type            10840 non-null  object 
 8   Price           10841 non-null  object 
 9   Content Rating  10841 non-null  object 
 10  Genres          10840 non-null  object 
 11  Last Updated    10841 non-null  object 
 12  Current Ver     10833 non-null  object 
 13  Android Ver     10839 non-null  object 
dtypes: float64(1), int64(2), object(11)
memory usage: 1.2+ MB


In [229]:
df.Size.dtype

dtype('O')

In [230]:
df.Size = df.Size.astype(float)

In [231]:
df.Size.dtype

dtype('float64')

In [232]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  float64
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  object 
 7   Type            10840 non-null  object 
 8   Price           10841 non-null  object 
 9   Content Rating  10841 non-null  object 
 10  Genres          10840 non-null  object 
 11  Last Updated    10841 non-null  object 
 12  Current Ver     10833 non-null  object 
 13  Android Ver     10839 non-null  object 
dtypes: float64(2), int64(2), object(10)
memory usage: 1.2+ MB


In [233]:
df.Installs.unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
       '10+', '1+', '5+', '0+', '0'], dtype=object)

In [234]:
for i in range(len(df.Installs)):
    df.Installs[i] = re.sub(",|\+","",str(df.Installs[i]))

C:\Users\shoba\AppData\Local\Temp/ipykernel_5920/3661904576.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.Installs[i] = re.sub(",|\+","",str(df.Installs[i]))


In [235]:
df.Installs.unique()

array(['10000', '500000', '5000000', '50000000', '100000', '50000',
       '1000000', '10000000', '5000', '100000000', '1000000000', '1000',
       '500000000', '50', '100', '500', '10', '1', '5', '0'], dtype=object)

In [236]:
df.Installs.dtype

dtype('O')

In [237]:
df.Installs = df.Installs.astype(np.int64)

In [238]:
df.Installs.dtype

dtype('int64')

In [239]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  float64
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  int64  
 7   Type            10840 non-null  object 
 8   Price           10841 non-null  object 
 9   Content Rating  10841 non-null  object 
 10  Genres          10840 non-null  object 
 11  Last Updated    10841 non-null  object 
 12  Current Ver     10833 non-null  object 
 13  Android Ver     10839 non-null  object 
dtypes: float64(2), int64(3), object(9)
memory usage: 1.2+ MB


In [240]:
df.Type.unique()

array(['Free', 'Paid', nan], dtype=object)

In [241]:
df[df.Type.isnull()]

App Category    Rating  Reviews  Size  Memory  \
9148  Command & Conquer: Rivals   FAMILY  4.191513        0   0.0       0   

      Installs Type Price Content Rating    Genres   Last Updated  \
9148         0  NaN     0   Everyone 10+  Strategy  June 28, 2018   

             Current Ver         Android Ver  
9148  Varies with device  Varies with device

In [242]:
df.Type.fillna('Free',inplace=True)

In [243]:
df.Type.value_counts()

Free    10041
Paid      800
Name: Type, dtype: int64

In [244]:
df.Price.unique()

array(['0', '$4.99', '$3.99', '$6.99', '$1.49', '$2.99', '$7.99', '$5.99',
       '$3.49', '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49',
       '$10.00', '$24.99', '$11.99', '$79.99', '$16.99', '$14.99',
       '$1.00', '$29.99', '$12.99', '$2.49', '$10.99', '$1.50', '$19.99',
       '$15.99', '$33.99', '$74.99', '$39.99', '$3.95', '$4.49', '$1.70',
       '$8.99', '$2.00', '$3.88', '$25.99', '$399.99', '$17.99',
       '$400.00', '$3.02', '$1.76', '$4.84', '$4.77', '$1.61', '$2.50',
       '$1.59', '$6.49', '$1.29', '$5.00', '$13.99', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$19.90', '$8.49', '$1.75',
       '$14.00', '$4.85', '$46.99', '$109.99', '$154.99', '$3.08',
       '$2.59', '$4.80', '$1.96', '$19.40', '$3.90', '$4.59', '$15.46',
       '$3.04', '$4.29', '$2.60', '$3.28', '$4.60', '$28.99', '$2.95',
       '$2.90', '$1.97', '$200.00', '$89.99', '$2.56', '$30.99', '$3.61',
       '$394.99', '$1.26', '$1.20', '$1.04'], dtype=object)

In [245]:
df.Price.value_counts()

0         10041
$0.99       148
$2.99       129
$1.99        73
$4.99        72
          ...  
$19.90        1
$1.75         1
$14.00        1
$4.85         1
$1.04         1
Name: Price, Length: 92, dtype: int64

In [247]:
for i in range(len(df.Price)):
    if df.Price[i] == '0':
        pass
    else:
        df.Price[i] = re.sub("\$","",str(df.Price[i]))

C:\Users\shoba\AppData\Local\Temp/ipykernel_5920/843572742.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.Price[i] = re.sub("\$","",str(df.Price[i]))


In [248]:
df.Price.value_counts()

0        10041
0.99       148
2.99       129
1.99        73
4.99        72
         ...  
19.90        1
1.75         1
14.00        1
4.85         1
1.04         1
Name: Price, Length: 92, dtype: int64

In [249]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  float64
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  int64  
 7   Type            10841 non-null  object 
 8   Price           10841 non-null  object 
 9   Content Rating  10841 non-null  object 
 10  Genres          10840 non-null  object 
 11  Last Updated    10841 non-null  object 
 12  Current Ver     10833 non-null  object 
 13  Android Ver     10839 non-null  object 
dtypes: float64(2), int64(3), object(9)
memory usage: 1.2+ MB


In [305]:
df.rename(columns={'Content Rating' : "Content_rating",'Last Updated': "Last_updated" , 'Current Ver': 'Current_version', "Android Ver" : "Android_version"}, inplace=True)

In [253]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  float64
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  int64  
 7   Type            10841 non-null  object 
 8   Price           10841 non-null  object 
 9   Content_rating  10841 non-null  object 
 10  Genres          10840 non-null  object 
 11  Last_updated    10841 non-null  object 
 12  Current Ver     10833 non-null  object 
 13  Android Ver     10839 non-null  object 
dtypes: float64(2), int64(3), object(9)
memory usage: 1.2+ MB


In [254]:
df.Content_rating.unique()

array(['Everyone', 'Teen', 'Everyone 10+', 'Mature 17+',
       'Adults only 18+', 'Unrated'], dtype=object)

In [255]:
df.Genres.unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

In [256]:
df[df.Genres.isnull()]

App     Category  Rating  Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame  PHOTOGRAPHY     1.9       19   

       Size  Memory  Installs  Type Price Content_rating Genres  \
10472   3.0       2      1000  Free     0       Everyone    NaN   

            Last_updated Current Ver Android Ver  
10472  February 11, 2018      1.0.19  4.0 and up

In [258]:
df.Genres.fillna('PHOTOGRAPHY', inplace=True)

In [259]:
df.Genres.isna().sum()

0

In [260]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  float64
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  int64  
 7   Type            10841 non-null  object 
 8   Price           10841 non-null  object 
 9   Content_rating  10841 non-null  object 
 10  Genres          10841 non-null  object 
 11  Last_updated    10841 non-null  object 
 12  Current Ver     10833 non-null  object 
 13  Android Ver     10839 non-null  object 
dtypes: float64(2), int64(3), object(9)
memory usage: 1.2+ MB


In [261]:
df.Last_updated.unique()

array(['January 7, 2018', 'January 15, 2018', 'August 1, 2018', ...,
       'January 20, 2014', 'February 16, 2014', 'March 23, 2014'],
      dtype=object)

In [276]:
y = []
m = []
d = []
for i in range(len(df.Last_updated)):
    y.append(df.Last_updated[i].split()[2])
    d.append(df.Last_updated[i].split()[1][0])
    m.append(df.Last_updated[i].split()[0])


In [277]:
d

['7',
 '1',
 '1',
 '8',
 '2',
 '2',
 '2',
 '1',
 '2',
 '3',
 '2',
 '3',
 '2',
 '2',
 '3',
 '6',
 '3',
 '7',
 '3',
 '3',
 '2',
 '2',
 '1',
 '7',
 '7',
 '2',
 '1',
 '2',
 '1',
 '2',
 '1',
 '3',
 '1',
 '6',
 '2',
 '1',
 '2',
 '1',
 '2',
 '3',
 '2',
 '2',
 '3',
 '1',
 '3',
 '3',
 '1',
 '1',
 '2',
 '1',
 '2',
 '2',
 '3',
 '2',
 '3',
 '1',
 '2',
 '2',
 '2',
 '3',
 '8',
 '3',
 '2',
 '1',
 '3',
 '3',
 '1',
 '3',
 '2',
 '1',
 '2',
 '2',
 '1',
 '2',
 '2',
 '1',
 '8',
 '4',
 '2',
 '1',
 '2',
 '2',
 '3',
 '2',
 '2',
 '4',
 '2',
 '2',
 '3',
 '1',
 '1',
 '3',
 '2',
 '2',
 '2',
 '2',
 '2',
 '2',
 '2',
 '9',
 '1',
 '1',
 '2',
 '2',
 '2',
 '1',
 '2',
 '5',
 '2',
 '1',
 '2',
 '2',
 '2',
 '2',
 '9',
 '1',
 '5',
 '3',
 '2',
 '2',
 '1',
 '2',
 '2',
 '2',
 '1',
 '1',
 '1',
 '2',
 '7',
 '8',
 '1',
 '2',
 '1',
 '1',
 '2',
 '5',
 '3',
 '7',
 '2',
 '1',
 '3',
 '3',
 '2',
 '2',
 '1',
 '1',
 '2',
 '2',
 '6',
 '2',
 '2',
 '1',
 '3',
 '2',
 '1',
 '1',
 '4',
 '2',
 '2',
 '2',
 '1',
 '2',
 '2',
 '1',
 '2',
 '4',
 '2'

In [267]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Memory', 'Installs',
       'Type', 'Price', 'Content_rating', 'Genres', 'Last_updated',
       'Current Ver', 'Android Ver'],
      dtype='object')

In [270]:
df.insert(12,'Year', y)

In [281]:
df.insert(13,'Month', m)
df.insert(14,'Date', d)

In [272]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  float64
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  int64  
 7   Type            10841 non-null  object 
 8   Price           10841 non-null  object 
 9   Content_rating  10841 non-null  object 
 10  Genres          10841 non-null  object 
 11  Last_updated    10841 non-null  object 
 12  Year            10841 non-null  object 
 13  Month           10841 non-null  object 
 14  Date            10841 non-null  object 
 15  Current Ver     10833 non-null  object 
 16  Android Ver     10839 non-null  object 
dtypes: float64(2), int64(3), object

In [282]:
df.Month.value_counts()

July         3163
August       1594
June         1273
May           978
March         667
April         616
February      534
January       491
December      426
October       398
November      387
September     314
Name: Month, dtype: int64

In [290]:
df.Month.replace(["January", 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'], [i for i in range(12)],inplace=True)

In [291]:
df.Month.dtype

dtype('int64')

In [292]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  float64
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  int64  
 7   Type            10841 non-null  object 
 8   Price           10841 non-null  object 
 9   Content_rating  10841 non-null  object 
 10  Genres          10841 non-null  object 
 11  Last_updated    10841 non-null  object 
 12  Year            10841 non-null  object 
 13  Month           10841 non-null  int64  
 14  Date            10841 non-null  object 
 15  Current Ver     10833 non-null  object 
 16  Android Ver     10839 non-null  object 
dtypes: float64(2), int64(4), object

In [293]:
df.drop("Last_updated", axis=1 , inplace=True)

In [295]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  float64
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  int64  
 7   Type            10841 non-null  object 
 8   Price           10841 non-null  object 
 9   Content_rating  10841 non-null  object 
 10  Genres          10841 non-null  object 
 11  Year            10841 non-null  object 
 12  Month           10841 non-null  int64  
 13  Date            10841 non-null  object 
 14  Current Ver     10833 non-null  object 
 15  Android Ver     10839 non-null  object 
dtypes: float64(2), int64(4), object(10)
memory usage: 1.3+ MB


In [296]:
df.Year.dtype

dtype('O')

In [297]:
df.Year = df.Year.astype(np.int64)

In [298]:
df.Date = df.Date.astype(np.int64)

In [299]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  float64
 5   Memory          10841 non-null  int64  
 6   Installs        10841 non-null  int64  
 7   Type            10841 non-null  object 
 8   Price           10841 non-null  object 
 9   Content_rating  10841 non-null  object 
 10  Genres          10841 non-null  object 
 11  Year            10841 non-null  int64  
 12  Month           10841 non-null  int64  
 13  Date            10841 non-null  int64  
 14  Current Ver     10833 non-null  object 
 15  Android Ver     10839 non-null  object 
dtypes: float64(2), int64(6), object(8)
memory usage: 1.3+ MB


In [306]:
df.Current_version.isnull().sum()

8

In [309]:
df[df.Current_version.isnull()]

App             Category    Rating  \
15     Learn To Draw Kawaii Characters       ART_AND_DESIGN  3.200000   
1553              Market Update Helper   LIBRARIES_AND_DEMO  4.100000   
6322            Virtual DJ Sound Mixer                TOOLS  4.200000   
6803                         BT Master               FAMILY  4.191513   
7333                       Dots puzzle               FAMILY  4.000000   
7407                   Calculate My IQ               FAMILY  4.191513   
7730                            UFO-CQ                TOOLS  4.191513   
10342                   La Fe de Jesus  BOOKS_AND_REFERENCE  4.191513   

       Reviews   Size  Memory  Installs  Type Price Content_rating  \
15          55    2.7       2      5000  Free     0       Everyone   
1553     20145   11.0       1   1000000  Free     0       Everyone   
6322      4010    8.7       2    500000  Free     0       Everyone   
6803         0  222.0       1       100  Free     0       Everyone   
7333       179   14.0       2     50000  Paid  0.99       Everyone   
7407        44    7.2       2     10000  Free     0       Everyone   
7730         1  237.0       1        10  Paid  0.99       Everyone   
10342        8  658.0       1      1000  Free     0       Everyone   

                  Genres  Year  Month  Date Current_version Android_ver  
15          Art & Design  2018      5     6             NaN  4.2 and up  
1553    Libraries & Demo  2013      1     1             NaN  1.5 and up  
6322               Tools  2017      4     1             NaN  4.0 and up  
6803           Education  2016     10     6             NaN  1.6 and up  
7333              Puzzle  2018      3     1             NaN  4.0 and up  
7407       Entertainment  2017      3     3             NaN  2.3 and up  
7730               Tools  2016      6     4             NaN  2.0 and up  
10342  Books & Reference  2017      0     3             NaN  3.0 and up